lib/ClangImporter/ClangImporter.cpp

```diff
@@ -3075,6 +3075,18 @@ ClangModuleUnit::ClangModuleUnit(ModuleDecl &M,
                                  const clang::Module *clangModule)
   : LoadedFile(FileUnitKind::ClangModule, M), owner(owner),
     clangModule(clangModule) {
+  // Capture the file metadata before it goes away.
+  if (clangModule)
+    ASTSourceDescriptor = {*clangModule};
+}
+
+Optional<clang::ExternalASTSource::ASTSourceDescriptor>
+ClangModuleUnit::getASTSourceDescriptor() const {
+  if (clangModule) {
+    assert(ASTSourceDescriptor.getModuleOrNull() == clangModule);
+    return ASTSourceDescriptor;
```

Comment:
This seems off to me. Isn't this the exact issue you're trying to fix with things going stale?

Reply:
What goes stale is the `ModuleFile *ASTFile` member of `clang::Module`. The `clang::Module` itself remains usable (except for that member) since it isn't owned by the `ModuleManager`.

Reply:
I still don't get it. Don't a clang::Module and all the clang::Decls that get deserialized point to data inside the ASTFile?

Reply:
You know the deserialization code much better than I do, I can only speculate for now. There are two possible explanations that I can imagine:
- all the Decl data is copied into the ASTContext's BumpPtrAllocator at deserialization time and any strings are internalized into StringMaps and such without keeping any StringRefs around. For example the source location information is safe because the source manager is separate from the module manager and the files stay around.
- or, none of the data is needed any more in Sema because it has already been lowered into Swift AST nodes. (Except for the debug info generation much later that wants the name and hash of the PCM).

I don't have a good explanation for that yet. I do know that the compilation comes out asan-clean all the way through IRGen.

Reply:
Since the ASTFile is a packed Bitstream, I wouldn't be that surprised if we don't need any pointers into the data in the compiler.

Reply:
Okay. The Swift modules definitely don't work that way but maybe the Clang ones do.